radareorg/cutter
# Rewrite layout management code

**Your checklist for this pull request**

**WARNING** :exclamation: 

This branch performs one way upgrade of settings and layout. If you have some complex layout that you want to keep using in the old Cutter backup your cutter settings.


**Detailed description**

Can be reviewed. Do not merge into 1.10.3 !

* Separate all layout state into a structure making it clearer when and where it's saved, loaded, switched.
* As part of layout state introduce dock view properties which is QVariant dictionary that can be used for easily saving single dock config. Makers of new widgets should consider if an option belongs to global settings or the new widget/layout specific properties. The new mechanism is suitable for situations where user might want to open two copies of a  widget with different view settings. For example two hexwidgets one showing memory as hex bytes, other as floating point numbers. To use the new mechanism developer needs to override @CutterDockWidget::serializeViewProperties@ and @CutterDockWidget::deSerializeViewProperties@   methods. 
* Allow saving multiple layout profiles in addition to builtin Default and Debug. The logic is that "Default" and "[default]Debug" always contain the last layout in corresponding mode. When user loads new profile or modifies the current layout it becomes the new "Default". Saved layouts are never automatically modified.
* Window size is separated from layout. It is loaded when opening Cutter. Switching layout doesn't touch the window size. It's possible to imagine that for some elaborate layouts window size is important but in most cases changing window size when switching layout would be unexpected. To be safe I kept the the window geometry in layout state but it isn't loaded. It will be easier to ignore or remove it later, than making it out of nothing in case it turns out to be necessary. Most examples show restoring state(dock placement) together with geometry(main window size and placement).
* Remove the custom mechanism for synchronizing widget visibility with action checkboxes in windows menu. Qt has it builtin, no need to reinvent it poorly. The old mechanism tried to deal with multiple widgets of the same kind but it only increased complexity and wasn't actually used.
* Code attempts to put new docks (new extra, plugins, or after Cutter upgrade) in the biggest area with existing disassembly, graph, hex widget. From my observations in https://github.com/radareorg/cutter/issues/694#issuecomment-621345529 and later it seems that any attempts to put something on the side may potentially result in broken layout. 

When I started working on this I tried to visualize layout related code structure. 
before:
![before](https://user-images.githubusercontent.com/7101031/80803229-62716d00-8bba-11ea-9ccc-1cd4da972e57.png)
after:
![after](https://user-images.githubusercontent.com/7101031/80803239-669d8a80-8bba-11ea-8978-5ae1fc42609e.png)


**Test plan (required)**

* Clean start without upgrade
  * remove cutter config
  * start cutter 
  * make sure default layout is reasonable
  * start debuging make sure layout is reasonable
* layout upgrade a)
  * remove cutter config
  * open 1.10.3 cutter
  * make minor modifications to layout
  * close 1.10.3 cutter
  * open new cutter make sure sure layout is reasonable
  * start debugging make sure layout is reasonable
* layout upgrade b)
  * remove cutter config
  * open 1.10.3 cutter
  * start debugging 
  * make some modification to layout 
  * stop debugging
  * close 1.10.3 cutter
  * open new cutter make sure sure layout is reasonable
  * start debugging make sure layout is reasonable
* saving layout modifications
  * open cutter modify layout
  * restart cutter, make sure layout is preserved
  * start debugging
  * modify debug layout
  * stop debugging
  * start debugging
  * make sure debug layout is preserved
  * restart cutter
  * start debugging 
  * make sure debug layout is preserved
  * stop debugging
* resetting layout (do after previous step)
  * be in non debug mode
  * reset layout
  * make sure it matches initial layout
  * start debugging
  * make sure debug layout isn't reset
  * reset layout
  * make sure you see the initial debug layout
  * modify both debug and normal layout again
  * reset settings
  * make sure both layouts were reset
* view property saving
  * Open multiple memory widgets(disassembly, hex or graph). At least two copies for one of them and place them side by side.
  * synchronize some of them, desynchronize others
  * restart cutter
  * make sure all the the memory widgets are opened and synchronization state is correctly restored

* It would be useful if reviewer and maybe more people tried a bunch of random things to see if some sequence of operations or specific layout can cause layout to break.
 
**Closing issues**

Not closing #694 , that should be done after finishing second part and implementing layout management UI.
Closes #1921
Maybe closes #1737 . Doesn't do exactly what was asked there, but in my opinion solves the problem in a somewhat acceptable way.


**Followup tasks**
* The new mechanism for saving view properties could be used to easily implement  #1704
* Better UI for managing layouts
* view properties could be used for saving some of hex widget config
* update Plugins to use the new API for dock action
* Overview widget showing/hiding. Looks there are some leftover code trying to do more than basic checkbox to show, but it doesn't seem to be working in current master or this PR. #2209
